### DISCUSSION TITLE
Multi Turbine using Coordinates and Modifying rotor diameter

### DISCUSSION THREAD

**Discussion by donghyeoncheon** [NatLabRockies/floris#718] [Q&A] [answered]
I'm planning to conduct a Wake Steering experiment by modeling 13 wind turbines using actual coordinates, and I'm looking for a way to apply these coordinates to an existing optimization code. Additionally, I'd like to know how to adjust the rotor diameter in the gch.yml file. When I did this in MATLAB, I used coordinate sampling for the model_type, rotor diameter, and layout_x, y variables. However, I'm unsure how to do this in Python. Please help!"

```python
import matplotlib.pyplot as plt
import numpy as np

from floris.tools import FlorisInterface
from floris.tools.layout_functions import visualize_layout
from floris.tools.optimization.yaw_optimization.yaw_optimizer_sr import YawOptimizationSR

import floris.tools.visualization as wakeviz


# Declare a FLORIS interface
fi = FlorisInterface("examples\inputs\gch.yaml")

# d = 40.0
#rotor_diameter = 1.0
fi.reinitialize(layout_x=[882.51,883.71,884.82,886.34,888.48,888.07,886.76,885.32,887.92,889.53,890.96,892.56,890.73], 
                layout_y=[1758.3,1755.87,1753.26,1751.23,1752.14,1755.57,1758.35,1760.55,1761.95,1759.4,1757.21,1754.7,1753.35],
                wind_directions=[270.0],
                wind_speeds=[7.4])
fi.calculate_wake()

# Give turbines specific names
turbine_names = ['T01', 'T02','T03','T04','T05', 'T06','T07', 'T08','T09','T10','T11', 'T12','T13']

# Plot the flow
fig, ax = plt.subplots(figsize=(8,8))
horizontal_plane = fi.calculate_horizontal_plane(
    x_resolution=1750,
    y_resolution=890,
    height=90.0
)
wakeviz.visualize_cut_plane(horizontal_plane, ax=ax, title="Horizontal")

# Plot the turnine names
visualize_layout(
    fi,
    ax=ax,
    show_wake_lines=False,
    turbine_names=turbine_names
)

plt.show()

yaw_opt = YawOptimizationSR(fi, verify_convergence=True)
df_opt = yaw_opt.optimize()

print("Optimization results:")
print(df_opt)

for t in range(13):
    df_opt['t%d' % t] = df_opt.yaw_angles_opt.apply(lambda x: x[t])
    
fig, axarr = plt.subplots(
    nrows=4,
    ncols=4,
    sharex=True,
    sharey=True,
    figsize=(10, 8)
)
jj = 0
for ii, ws in enumerate(fi.floris.flow_field.wind_speeds):
    xi = np.remainder(ii,13)
    if ((ii > 0) & (xi == 0)):
        jj += 1
    ax = axarr[np.remainder(ii, 13)][jj]
    ids = (df_opt.wind_speed == ws)
    wd = df_opt.loc[ids, "wind_direction"]
    for t in range(13):
        yaw_opt = df_opt.loc[ids, "t{:d}".format(t)]
        ax.plot(wd, yaw_opt, label='Turbine {:d}'.format(t))
    ax.set_title("Wind speed = {:.1f} m/s".format(ws), size=10)
    if ((ii == 0) & (jj == 0)):
        ax.legend()
    ax.grid(True)
    if jj == 0:
        ax.set_ylabel('Yaw angle (deg)', size=10)
    if xi == 3:
        axarr[xi][jj].set_xlabel('Wind Direction (deg)', size=10)

    plt.tight_layout()
    
fig, axarr = plt.subplots(
    nrows=5,
    ncols=4,
    sharex=True,
    sharey=True,
    figsize=(10, 8)
)
jj = 0
for ii, ws in enumerate(fi.floris.flow_field.wind_speeds):
    xi = np.remainder(ii, 5)
    if ((ii > 0) & (xi == 0)):
        jj += 1
    ax = axarr[np.remainder(ii, 5)][jj]
    ids = (df_opt.wind_speed == ws)
    wd = df_opt.loc[ids, "wind_direction"]
    power_baseline = df_opt.loc[ids, "farm_power_baseline"]
    power_opt = df_opt.loc[ids, "farm_power_opt"]
    ax.plot(wd, power_baseline / 1e6, color='k', label='Baseline')
    ax.plot(wd, power_opt / 1e6, color='r', label='Optimized')
    ax.set_title("Wind speed = {:.1f} m/s".format(ws), size=10)
    ax.set_ylim([0.0, 16.0])
    if ((ii == 0) & (jj == 0)):
        ax.legend()
    ax.grid(True)
    if jj == 0:
        ax.set_ylabel('Farm Power (MW)', size=10)
    if xi == 3:
        axarr[xi][jj].set_xlabel('Wind Direction (deg)', size=10)

    plt.tight_layout()

    plt.show()
```

At the point is.. Modeling 13 turbines using coordinates and optimization using for wake steering and check the power efficiency... 

![image](https://github.com/NREL/floris/assets/58211121/17a6ade9-34b8-4d7c-a006-95fae1c1bc0c)
this problem..too..!

**Comment by paulf81** [comment]
hi @donghyeoncheon , thank you for your question!

To your question on, the rotor diameter is specified in the turbine model file.  Since you are using the gch.yaml input file you can see it specifies the nrel_5mw turbine model, which is located:
floris/turbine_library/nrel_5MW.yaml

On your next question I was wondering if adapting on of the existing optimization examples would work for you, say examples/14_compare_yaw_optimizers.py?

Paul

**Reply by donghyeoncheon** [reply]
Thank you for reply me @paulf81..! I'll try again!

**Reply by rafmudaf** [reply]
@paulf81 @donghyeoncheon it's not recommended to change the rotor diameter in the turbine library directly. Instead, we currently have to do this through the input script when using one of the internal turbines. See below as suggested in https://github.com/NREL/floris/discussions/595#discussioncomment-5177339:

```python
fi = FlorisInterface("inputs/gch.yaml")
print(fi.floris.farm.rotor_diameters)

### Update rotor diameters manually

# Copy properties of first turbine
turbine_dict = fi.floris.farm.turbine_definitions[0]

# Change rotor diameter
turbine_dict["rotor_diameter"] = 107.0

# Copy these turbine properties over to make a list with length equal to the number of wind turbines in the farm
turbine_dicts = [turbine_dict] * fi.floris.farm.n_turbines

# Update FLORIS object with these new turbine definitions
fi.reinitialize(turbine_type=turbine_dicts)

print(fi.floris.farm.rotor_diameters)
```

**Comment by rafmudaf** [comment]
@donghyeoncheon Can you describe how you expect the layout of your farm to look? You could plot your coordinates (`layout_x` and `layout_y`) in MATLAB (or another tool such as Excel) and share the plot here.

**Reply by donghyeoncheon** [reply]
In case of MATLAB result.. 
![image](https://github.com/NREL/floris/assets/58211121/eda00491-92ad-4829-b2dc-951a029a99f7)
i made a result like this. but my coordinate is too near per turbine. so i reduced rotor diameter in matlab. and this is my coordinate of model on the matlab 
![image](https://github.com/NREL/floris/assets/58211121/c7c2891b-d04a-4c59-a5e5-3f7ad13895e6)

**Reply by rafmudaf** [reply]
Oh so are you trying to reduce the rotor diameter in floris?

**Reply by donghyeoncheon** [reply]
It is better to proceed without reducing the Rotor Diameter. Still, the wind farm generator coordinates are too close, so the location will appear correctly on the plot only when I reduce it. That Wind farm consisted of 1.5 MW * 7 , 0.75 MW *6, 15MW..

**Reply by rafmudaf** [reply]
Sorry, I think I don't fully understand your situation.
Can you scale the layout to reflect the actual turbine diameter?

**Reply by donghyeoncheon** [reply]
![image](https://github.com/NREL/floris/assets/58211121/f0ff74c3-d1be-49c1-873c-a6743281fbfb)
well.. this is Row data.. and you can see the coordinate on Matlab.. that is scaled 1:100. I tried to modify inputs and turbine library on FLORIS that didn't fully work.. And When the coordinates are entered into the code to check the turbine locations, the positions of each turbine are so close that if not zoomed in on the plot, the results overlap and are displayed together. 
![image](https://github.com/NREL/floris/assets/58211121/ba4cb53e-32a8-49cc-8afc-90c7033ea790)
I am considering how to adjust the size of the wind direction 

I apologize for being disorganized. I have two issues. First, do I need to create a new model type for modeling using coordinates? And second, how do I adjust the size of the wind direction on the plot when I input coordinates into the sample code?"

**Reply by rafmudaf** [reply, marked as answer]
@donghyeoncheon You can scale the inputs in FLORIS similar to what you did in Matlab:

```python
layout_x = 100 * np.array([882.51,883.71,884.82,886.34,888.48,888.07,886.76,885.32,887.92,889.53,890.96,892.56,890.73])
layout_y = 100 * np.array([1758.3,1755.87,1753.26,1751.23,1752.14,1755.57,1758.35,1760.55,1761.95,1759.4,1757.21,1754.7,1753.35])
fi.reinitialize(
    layout_x=layout_x, 
    layout_y=layout_y,
    wind_directions=[270.0],
    wind_speeds=[7.4]
)
fi.calculate_wake()
```

However, I'm wondering how you got these coordinates. Did you manually convert from lat / long to another form? If not, please describe how you got the x and y values.

**Reply by donghyeoncheon** [reply]
Thank you for replying! I'll try that again :D. 
I received these coordinates from a managed energy company in Korea.
I only examined the distribution chart using those coordinates and didn't convert the data in any way. 
However, I did verify the authenticity of the information.
